### PR TITLE
Fix get failed/errored transactions sink

### PR DESCRIPTION
**Pull request by Daniel-VDM**
## Issue

Currently, one can only get failed/errored transactions if one knows the leader's IP address. This PR checks for failed transactions in the transaction pool, this any node can report failed or errored transactions.

### Unit Test Coverage

Before:

```
newNodeList:  [ECDSA: one1qqqrvvpcxu6rwvfnxc6ngvecx5mrgvf3254yj4, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qq6n2de3xuuryven8qcnqvfcxuurwd3s837n5r, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qqcnjv3kxqcnydfcxc6nyd3kxg6rqvpe5srx9h, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qqqrgvp5xy6nxwf5x5mngve4xsmnvdfhekh0gx, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qqen2ve5xvengvekxuerzdpjxvmnyd33m0urmm, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000]
PASS
coverage: 24.5% of statements
ok  	github.com/harmony-one/harmony/core	6.960s
```

```
PASS
coverage: 12.9% of statements
ok  	github.com/harmony-one/harmony/node	2.239s
```

After:

```
newNodeList:  [ECDSA: one1qqqrvvpcxu6rwvfnxc6ngvecx5mrgvf3254yj4, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qq6n2de3xuuryven8qcnqvfcxuurwd3s837n5r, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qqcnjv3kxqcnydfcxc6nyd3kxg6rqvpe5srx9h, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qqqrgvp5xy6nxwf5x5mngve4xsmnvdfhekh0gx, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 ECDSA: one1qqen2ve5xvengvekxuerzdpjxvmnyd33m0urmm, BLS: 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000]
PASS
coverage: 24.9% of statements
ok  	github.com/harmony-one/harmony/core	6.774s
```

```
PASS
coverage: 13.3% of statements
ok  	github.com/harmony-one/harmony/node	1.933s
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    **NO**

### CODE REVIEW

File: core/tx_pool.go [hunk 1]
```diff
@@ -1163,6 +1184,11 @@ func (pool *TxPool) demoteUnexecutables() {
 		for _, tx := range list.Forward(nonce) {
 			hash := tx.Hash()
 			logger.Warn().Str("hash", hash.Hex()).Msg("Removed old pending transaction")
+			erroredTxns = append(erroredTxns, types.RPCTransactionError{
+				TxHashID:             hash.Hex(),
+				TimestampOfRejection: time.Now().Unix(),
+				ErrMessage:           "old pending transaction",
```
Comment (rlan35) [inline core/tx_pool.go]: more informative message be can returned. Just adding the nonces.

File: core/tx_pool.go [hunk 2]
```diff
@@ -1171,20 +1197,35 @@ func (pool *TxPool) demoteUnexecutables() {
 		for _, tx := range drops {
 			hash := tx.Hash()
 			logger.Warn().Str("hash", hash.Hex()).Msg("Removed unpayable pending transaction")
+			erroredTxns = append(erroredTxns, types.RPCTransactionError{
+				TxHashID:             hash.Hex(),
+				TimestampOfRejection: time.Now().Unix(),
+				ErrMessage:           "unpayable pending transaction",
+			})
```
Comment (rlan35) [inline core/tx_pool.go]: Same here, add the balances or gas cost. Basically we want the user to get least confused.

File: core/tx_pool.go [hunk 3]
```diff
@@ -860,6 +863,13 @@ func (pool *TxPool) addTx(tx *types.Transaction, local bool) error {
 	// Try to inject the transaction and update any state
 	replace, err := pool.add(tx, local)
 	if err != nil {
+		pool.txnErrorSink([]types.RPCTransactionError{
+			{
```
Comment (fxfactorial) [inline core/tx_pool.go]: did gofmt accept this?

File: node/node.go [hunk 4]
```diff
@@ -490,7 +490,15 @@ func New(host p2p.Host, consensusObj *consensus.Consensus, chainDBFactory shardc
 		node.ConfirmedBlockChannel = make(chan *types.Block)
 		node.BeaconBlockChannel = make(chan *types.Block)
 		node.recentTxsStats = make(types.RecentTxsStats)
-		node.TxPool = core.NewTxPool(core.DefaultTxPoolConfig, node.Blockchain().Config(), blockchain)
+		node.TxPool = core.NewTxPool(core.DefaultTxPoolConfig, node.Blockchain().Config(), blockchain,
+			func(payload []types.RPCTransactionError) {
+				node.errorSink.Lock()
```
Comment (rlan35) [inline node/node.go]: The lock is fine if it's not frequently called. I don't think so.

File: core/tx_pool.go (hunk 3 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: Yep, passed Travis and I just ran go fmt on it.

File: core/tx_pool.go (hunk 1 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: Added informative msgs in latest

File: core/tx_pool.go (hunk 2 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: Added informative msgs in latest

File: node/node.go (hunk 4 above)
Comment (Daniel-VDM) [reply node/node.go]: Added a len check, in latest, just incase

File: core/tx_pool.go [hunk 5]
```diff
@@ -26,6 +26,8 @@ import (
 	"sync"
 	"time"
 
+	errors2 "github.com/pkg/errors"
```
Comment (fxfactorial) [inline core/tx_pool.go]: don't do this. pkg/errors is a superset of the stdlib errors, so you don't need to keep plain error and errors. just remove error and use pkg/errors without this error2 alias

Comment (Daniel-VDM) [reply core/tx_pool.go]: alright, changed it in latest

File: core/tx_pool.go [hunk 6]
```diff
@@ -601,42 +605,46 @@ func (pool *TxPool) local() map[common.Address]types.Transactions {
 func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
 	// Heuristic limit, reject transactions over 32KB to prevent DOS attacks
 	if tx.Size() > 32*1024 {
```
Comment (LeoHChen) [inline core/tx_pool.go]: Does this number still make sense?  Maybe open a follow up ticket for further investigation.

File: core/tx_pool.go [hunk 7]
```diff
@@ -601,42 +605,46 @@ func (pool *TxPool) local() map[common.Address]types.Transactions {
...
-		return ErrNonceTooLow
+		return errors.WithMessagef(ErrNonceTooLow, "transaction nonce is %d", tx.Nonce())
 	}
 	// Transactor should have enough funds to cover the costs
 	// cost == V + GP * GL
 	if pool.currentState.GetBalance(from).Cmp(tx.Cost()) < 0 {
-		return ErrInsufficientFunds
+		cost := new(big.Float).SetInt64(tx.Cost().Int64())
+		cost = cost.Mul(cost, new(big.Float).SetFloat64(1e-18)) // Cost is in Atto
+		return errors.WithMessagef(ErrInsufficientFunds, "transaction cost is %.18f ONE", cost)
```
Comment (LeoHChen) [inline core/tx_pool.go]: Does it make sense to use a different error message here than the previous one?

File: core/tx_pool.go [hunk 8]
```diff
@@ -601,42 +605,46 @@ func (pool *TxPool) local() map[common.Address]types.Transactions {
...
 	// Ensure the transaction adheres to nonce ordering
 	if pool.currentState.GetNonce(from) > tx.Nonce() {
-		return ErrNonceTooLow
+		return errors.WithMessagef(ErrNonceTooLow, "transaction nonce is %d", tx.Nonce())
 	}
 	// Transactor should have enough funds to cover the costs
 	// cost == V + GP * GL
 	if pool.currentState.GetBalance(from).Cmp(tx.Cost()) < 0 {
-		return ErrInsufficientFunds
+		cost := new(big.Float).SetInt64(tx.Cost().Int64())
```
Comment (rlan35) [inline core/tx_pool.go]: Float may have precision issue.

File: core/tx_pool.go (hunk 7 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: I suppose not, also float may give a slightly wrong cost. Will just return error.

File: core/tx_pool.go (hunk 6 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: Opened a ticket, #2154

File: core/tx_pool.go (hunk 8 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: Yep, extra msg is also not needed. Will just return the error.

**Pull request merged**